pkg/proc/native/threads_linux_arm.go

```diff
@@ -0,0 +1,247 @@
+package native
+
+import (
+	"bytes"
+	"debug/elf"
+	"encoding/binary"
+	"fmt"
+	"golang.org/x/arch/arm/armasm"
+	"math/bits"
+	"syscall"
+	"unsafe"
+
+	sys "golang.org/x/sys/unix"
+
+	"github.com/go-delve/delve/pkg/proc"
+	"github.com/go-delve/delve/pkg/proc/linutil"
+)
+
+func (t *nativeThread) fpRegisters() ([]proc.Register, []byte, error) {
+	var err error
+	var arm_fpregs linutil.ARM64PtraceFpRegs
+	t.dbp.execPtraceFunc(func() { arm_fpregs.Vregs, err = ptraceGetFpRegset(t.ID) })
+	fpregs := arm_fpregs.Decode()
+	if err != nil {
+		err = fmt.Errorf("could not get floating point registers: %v", err.Error())
+	}
+	return fpregs, arm_fpregs.Vregs, err
+}
+
+func (t *nativeThread) restoreRegisters(savedRegs proc.Registers) error {
+	sr := savedRegs.(*linutil.ARMRegisters)
+
+	var restoreRegistersErr error
+	t.dbp.execPtraceFunc(func() {
+		restoreRegistersErr = ptraceSetGRegs(t.ID, sr.Regs)
+		if restoreRegistersErr != syscall.Errno(0) {
+			return
+		}
+		if sr.Fpregset != nil {
+			iov := sys.Iovec{Base: &sr.Fpregset[0], Len: uint32(len(sr.Fpregset))}
+			_, _, restoreRegistersErr = syscall.Syscall6(syscall.SYS_PTRACE, sys.PTRACE_SETREGSET, uintptr(t.ID), uintptr(elf.NT_FPREGSET), uintptr(unsafe.Pointer(&iov)), 0, 0)
+		}
+	})
+	if restoreRegistersErr == syscall.Errno(0) {
+		restoreRegistersErr = nil
+	}
+	return restoreRegistersErr
+}
+
+func (t *nativeThread) singleStep() (err error) {
+	// Arm don't have ptrace singleStep implemented, so we use breakpoint to emulate it.
+	for {
+		regs, err := t.Registers()
+		if err != nil {
+			return err
+		}
+		resolvePCs := func() ([]uint64, []uint64, error) {
```

Comment:
I think this function would be easier to understand if it was defined outside of (*nativeThread).singleStep

Reply:
I will refactor it as member function of nativeThread.